### PR TITLE
[GCS] Monitor.py bug fix

### LINKED ISSUE #8715: [autoscaler] Cluster becomes unusable when job exits.
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):* 0.9dev

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

If we cannot run your script, we cannot fix your issue.

I ran with the default AWS autoscaler config (`python/ray/autoscaler/aws/example-full.yaml`), modified the minimum to 2 workers. Steps to reproduce:
1. SSH into head node.
2. Run a Ray script from bash (`while True: ray.get(foo.remote())` works)
3. Cancel Ray script with CTRL+C.

Then, in `/tmp/ray/session_latest/logs/monitor.err`, I get this:
```
2020-06-01 23:06:52,035 INFO autoscaler.py:619 -- StandardAutoscaler: 2/2 target nodes (0 pending) (2 updating) (bringup=True)
2020-06-01 23:06:52,035 INFO autoscaler.py:620 -- LoadMetrics: MostDelayedHeartbeats={'172.30.0.103': 0.1708686351776123}, NodeIdleSeconds=Min=0 Mean=0 Max=0, NumNodesConnected=1, NumNodesUsed=0.02, ResourceUsage=1.0/64.0 CPU, 0.0 GiB/178.57 GiB memory, 0.0/1.0 node:172.30.0.103, 0.0 GiB/55.73 GiB object_store_memory, TimeSinceLastHeartbeat=Min=0 Mean=0 Max=0
2020-06-01 23:06:56,727 INFO updater.py:257 -- NodeUpdater: i-0fe7dc6a0f3fd63be: Running uptime on 172.30.1.111...
2020-06-01 23:06:56,733 INFO updater.py:257 -- NodeUpdater: i-0a74d5ee1dadf1b74: Running uptime on 172.30.1.30...
Warning: Permanently added '172.30.1.111' (ECDSA) to the list of known hosts.^M
Warning: Permanently added '172.30.1.30' (ECDSA) to the list of known hosts.^M
Error in monitor loop
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/monitor.py", line 277, in run
    self._run()
  File "/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/monitor.py", line 235, in _run
    self.process_messages()
  File "/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/monitor.py", line 187, in process_messages
    message_handler(channel, data)
  File "/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/monitor.py", line 125, in xray_job_notification_handler
    job_data = gcs_entries.entries[0]
IndexError: list index (0) out of range
2020-06-01 23:06:56,765 ERROR autoscaler.py:645 -- StandardAutoscaler: kill_workers triggered
2020-06-01 23:06:56,828 INFO node_provider.py:353 -- AWSNodeProvider: terminating nodes ['i-0fe7dc6a0f3fd63be', 'i-0a74d5ee1dadf1b74'] (spot nodes cannot be stopped, only terminated)
2020-06-01 23:06:56,997 ERROR autoscaler.py:650 -- StandardAutoscaler: terminated 2 node(s)
2020-06-01 23:06:57,123 INFO commands.py:93 -- teardown_cluster: Keeping 2 nodes...
Monitor: Cleanup exception. Trying again...
mux_client_request_session: read from master failed: Broken pipe^M
Failed to connect to new control master^M
mux_client_request_session: read from master failed: Broken pipe^M
Failed to connect to new control master^M
2020-06-01 23:06:59,270 INFO commands.py:93 -- teardown_cluster: Keeping 2 nodes...
Monitor: Cleanup exception. Trying again...
2020-06-01 23:07:01,385 INFO commands.py:93 -- teardown_cluster: Keeping 2 nodes...
Monitor: Cleanup exception. Trying again...
2020-06-01 23:07:02,231 INFO updater.py:257 -- NodeUpdater: i-0a74d5ee1dadf1b74: Running uptime on 172.30.1.30...
ssh: connect to host 172.30.1.30 port 22: Connection refused^M
2020-06-01 23:07:02,243 INFO updater.py:257 -- NodeUpdater: i-0fe7dc6a0f3fd63be: Running uptime on 172.30.1.111...
ssh: connect to host 172.30.1.111 port 22: Connection refused^M
2020-06-01 23:07:03,513 INFO commands.py:93 -- teardown_cluster: Keeping 2 nodes...
Monitor: Cleanup exception. Trying again...
2020-06-01 23:07:05,622 INFO commands.py:93 -- teardown_cluster: Keeping 2 nodes...
Monitor: Cleanup exception. Trying again...
```

Other times I get the same Python exception and teardown messages, but not the infinitely repeating messages about "Cleanup exception". However, the messages just stop completely and no new worker nodes are started. I didn't confirm, but this might be if I wait for the worker nodes to come up completely before starting the job.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?
This fixes https://github.com/ray-project/ray/issues/8715

The new pub/sub mechanism wasn't compatible with monitor.py and it was broken. The issue turns out that we should use PubSubMessage instead of GCSEntry for some entries. 

## Related issue number

https://github.com/ray-project/ray/issues/8715

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)
   - [x] Manual testing

### CODE REVIEW

File: python/ray/monitor.py [hunk 1]
```diff
@@ -121,9 +121,7 @@ def xray_job_notification_handler(self, unused_channel, data):
             unused_channel: The message channel.
             data: The message data.
         """
-        gcs_entries = ray.gcs_utils.GcsEntry.FromString(data)
-        job_data = gcs_entries.entries[0]
-        message = ray.gcs_utils.JobTableData.FromString(job_data)
+        message = ray.gcs_utils.JobTableData.FromString(data)
```
Comment (ffbin) [inline python/ray/monitor.py:124]: pub_message = ray.gcs_utils.PubSubMessage.FromString(data)
        job_data = pub_message.data
        message = ray.gcs_utils.JobTableData.FromString(job_data)

**Comment by ffbin** [comment]
image

**Comment by rkooo567** [comment]
@ffbin It makes sense! So, for entries that use Redis accessor, we should use GCSEntry and for others it is PubSubMessage? Why didn't we make it work in the same way as before?

**Review by stephanie-wang — APPROVED** [review]
Thanks! Is there any way to test this?

**Comment by rkooo567** [comment]
@stephanie-wang I manually tested, and it worked. 

But, I think autoscaler tests should've caught this issue. I am not sure why it didn't do that.

**Comment by ffbin** [comment]
GcsEntry is used by redis module, PubSubMessage is used by gcs pub-sub. Currently, GCS pub sub is used between GCS server and GCS client. If python code needs to subscribe to GCS table, it needs to use PubSubMessage, thanks.

**Pull request merged**